### ISSUE TITLE
[REVIEW]: NeuralHydrology --- A Python library for Deep Learning research in hydrology

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#4050]
**Submitting author:** @kratzert (<a href="http://orcid.org/0000-0002-8897-7689">Frederik Kratzert</a>)
**Repository:** https://github.com/neuralhydrology/neuralhydrology
**Branch with paper.md** (empty if default branch): paper
**Version:** v.1.2.3
**Editor:** @elbeejay
**Reviewers:** @ammilten, @chuckaustin, @jhamman
**Archive:** 10.5281/zenodo.6326394

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/d3567e525e20638a8a27c8c6036e1bd7/status.svg)](https://joss.theoj.org/papers/d3567e525e20638a8a27c8c6036e1bd7)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/d3567e525e20638a8a27c8c6036e1bd7"><img src="https://joss.theoj.org/papers/d3567e525e20638a8a27c8c6036e1bd7/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/d3567e525e20638a8a27c8c6036e1bd7/status.svg)](https://joss.theoj.org/papers/d3567e525e20638a8a27c8c6036e1bd7)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/neuralhydrology/neuralhydrology" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@ammilten & @chuckaustin & @jhamman, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @elbeejay know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @ammilten
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/neuralhydrology/neuralhydrology">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/neuralhydrology/neuralhydrology">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@kratzert) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @chuckaustin
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/neuralhydrology/neuralhydrology">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/neuralhydrology/neuralhydrology">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@kratzert) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @jhamman
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/neuralhydrology/neuralhydrology">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/neuralhydrology/neuralhydrology">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@kratzert) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @ammilten, @chuckaustin, @jhamman it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
PDF failed to compile for issue #4050 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.21 s (692.2 files/s, 85033.4 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          60           2303           3770           5437
YAML                            18            246            217            807
reStructuredText                55            298            366            746
Jupyter Notebook                 6              0           2878            297
DOS Batch                        1              8              1             26
Markdown                         1              7              0             16
make                             1              4              7              9
-------------------------------------------------------------------------------
SUM:                           142           2866           7239           7338
-------------------------------------------------------------------------------


Statistical information for the repository '31a2108e77ad1a07e6e208aa' was
gathered on 2022/01/11.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Daniel Klotz                     2          1392            437            9.70
Frederik Kratzert               54         12028           2220           75.53
Grey Nearing                     1           259             14            1.45
Martijn Visser                   1             9             10            0.10
Martin Gauch                    32          1692            804           13.23

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Daniel Klotz               1028           73.9         10.1                9.05
Frederik Kratzert          8753           72.8          3.7                9.37
Grey Nearing                199           76.8         11.7                8.54
Martijn Visser                9          100.0          8.0               22.22
Martin Gauch               1521           89.9         11.6                8.68
```

**Comment by elbeejay** [comment]
@ammilten, @chuckaustin, @jhamman, thanks for agreeing to review this submission to JOSS. We are currently asking reviewers to try and complete their reviews in 6 weeks.

The JOSS review process is entirely open and transparent, and takes place on GitHub. **Review comments should be made as issues in the NeuralHydrology repository** (so open up a new issue [here](https://github.com/neuralhydrology/neuralhydrology/issues) with review comments), please link this review issue when doing so (by pasting `https://github.com/openjournals/joss-reviews/issues/4050` into the review issue you create).

For reference, here are links to the JOSS documentation that may be helpful as you conduct your reviews:

[JOSS Review Criteria](https://joss.readthedocs.io/en/latest/review_criteria.html)
[JOSS Review Checklist](https://joss.readthedocs.io/en/latest/review_checklist.html)

Please feel free to ping me (@elbeejay) if you have any questions/concerns. Thanks again for agreeing to review for JOSS.

**Comment by elbeejay** [comment]
@whedon check references from branch paper

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch paper
```

**Comment by elbeejay** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
Checking the BibTeX entries failed with the following error: 

 Failed to parse BibTeX entry: cite-key missing

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by elbeejay** [comment]
@kratzert - if looks like there's an issue with the bibliography file. I suspect it has to do with an "@" symbol in the citation:

> @article{kratzert@regional,
> author = {Kratzert, F. and Klotz, D. and Shalev, G. and Klambauer, G. and Hochreiter, S. and Nearing, G.},
> title = {Towards learning universal, regional, and local hydrological behaviors via machine learning applied to large-sample datasets},
> journal = {Hydrology and Earth System Sciences},
> volume = {23},
> year = {2019},
> number = {12},
> pages = {5089--5110},
> doi = {10.5194/hess-23-5089-2019}
> }

Can you remove the second "@" and update the bib file? Once you've done that please use the whedon commands like above to regenerate the paper and check the reference file.

**Comment by kratzert** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.04050/joss.04050/10.21105.joss.04050.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.04050/joss.04050/10.21105.joss.04050.pdf) :page_facing_up: :point_left:

**Comment by kratzert** [comment]
Sorry for the spam but I just saw that we had different spellings for the same authors across different bib entries, which led to an undesired format in the reference list. I fixed this and will query whedon again to compile the manuscript. Please consider the document below for the review.

@whedon generate pdf from branch paper

**Comment by kratzert** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by elbeejay** [comment]
@whedon check references from branch paper

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch paper
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.5194/hess-23-5089-2019 is OK
- 10.1029/2019WR026065 is OK
- 10.5194/hess-25-2045-2021 is OK
- 10.5194/hess-2021-154 is OK
- 10.5194/hess-2021-127 is OK

MISSING DOIs

- 10.2166/9781780401041 may be a valid DOI for title: Evolution of Water Supply Through the Millennia

INVALID DOIs

- https://doi.org/10.1029/2020WR028091 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1002/hyp.13805 is INVALID because of 'https://doi.org/' prefix
```

**Comment by kratzert** [comment]
@whedon check references from branch paper

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch paper
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.2166/9781780401041 is OK
- 10.5194/hess-23-5089-2019 is OK
- 10.1029/2019WR026065 is OK
- 10.5194/hess-25-2045-2021 is OK
- 10.5194/hess-2021-154 is OK
- 10.5194/hess-2021-127 is OK
- 10.1029/2020WR028091 is OK
- 10.1002/hyp.13805 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @jhamman, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @ammilten, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @chuckaustin, please update us on how your review is going (this is an automated reminder).

**Comment by elbeejay** [comment]
Hi all, just to check-in and clarify - that is a 2 week reminder above. As JOSS is still in a "[reduced service mode](https://blog.joss.theoj.org/2020/05/reopening-joss)" due to COVID, we are looking for your reviews to be complete within 6 weeks total, which would be 4 weeks from today (Feb. 22). 

We are of course understanding of any potential delays or personal issues which may prevent you from conducting your review within that time-span, but ask that you please let us know ahead of time here if your review is going to be delayed.

Otherwise thank you again for reviewing for JOSS, and reach out to me if you have any questions as you conduct your reviews.

**Comment by jhamman** [comment]
I'm grateful I was able to review this submission. I have a few comments relative to my review:

- The article indicates the package is installable via `pip` but upon inspection, the package has not been published on PyPI (or another package manager like conda-forge).  I've opened https://github.com/neuralhydrology/neuralhydrology/issues/66 to suggest this be done.
- There appears to be a missing citation under **Example Use Cases** in the article. See _Kratzert et al. 20xx_.
- The authors are encouraged to add a citation to the the following claim:
  > Users can easily train an ensemble of multiple models and average their predictions, which often greatly enhances the overall accuracy.

**Comment by ammilten** [comment]
I really enjoyed going through this code. Once I got the intro script working and figured out the structure of the code, I really started to appreciate the amount of effort the author put in to making deep learning accessible. The way the code uses .yml files to set the training, etc. settings makes it really easy to work with deep learning. 

Before publishing, the only thing I see that needs to be addressed is the missing citation in the paper (e.g. Kratzert et al. 20xx)

Once this is addressed I support accepting this submission.

A few suggested features for the future:
1. Python wrappers for reading/writing .yml files would be very useful.
2. Automated tests, if possible
3. Model analysis: Explainability (e.g. LIME/SHAP), feature importances, uncertainty quantification (e.g. Monte Carlo dropout). These would be lots of work but if you ever use them in your own research it would be cool to see them added to this library.

These are just suggestions and not necessary before acceptance. Congratulations to the authors for a high-quality, comprehensive piece of work.

Alex

**Comment by kratzert** [comment]
> I really enjoyed going through this code. Once I got the intro script working and figured out the structure of the code, I really started to appreciate the amount of effort the author put in to making deep learning accessible. The way the code uses .yml files to set the training, etc. settings makes it really easy to work with deep learning.
> 
> Before publishing, the only thing I see that needs to be addressed is the missing citation in the paper (e.g. Kratzert et al. 20xx)
> 
> Once this is addressed I support accepting this submission.
> 
> A few suggested features for the future:
> 
>     1. Python wrappers for reading/writing .yml files would be very useful.
> 
>     2. Automated tests, if possible
> 
>     3. Model analysis: Explainability (e.g. LIME/SHAP), feature importances, uncertainty quantification (e.g. Monte Carlo dropout). These would be lots of work but if you ever use them in your own research it would be cool to see them added to this library.
> 
> 
> These are just suggestions and not necessary before acceptance. Congratulations to the authors for a high-quality, comprehensive piece of work.
> 
> Alex

Thanks @ammilten for your time and your valuable feedback on our tutorials.

- The reference is fixed, see above (and I will invoke the bot to recompile the paper for this thread).

Regarding your suggestions:
- Not sure if this is what you mean, but we already have utilities for reading/writing .yml files. Most of it is related to our `Config` class (see [docs](https://neuralhydrology.readthedocs.io/en/latest/api/neuralhydrology.utils.config.html)), which we use everywhere in the code to parse the .yml config arguments. The class has a public API for reading yaml files (the object is initialized from the yaml), to [change/update](https://neuralhydrology.readthedocs.io/en/latest/api/neuralhydrology.utils.config.html#neuralhydrology.utils.config.Config.update_config) the config arguments (via the `.update_config()` method, and to [store](https://neuralhydrology.readthedocs.io/en/latest/api/neuralhydrology.utils.config.html#neuralhydrology.utils.config.Config.dump_config) the config to a yaml file (via the `dump_config()` method, which deals with translating Path objects to string and converting pandas Timestamps to strings). Furthermore, there is also a `as_dict()` method, in case anyone wants to work with the parsed yaml config as a Python dictionary. We also have the [create_config_files](https://neuralhydrology.readthedocs.io/en/latest/api/neuralhydrology.utils.configutils.html#neuralhydrology.utils.configutils.create_config_files) function that can be used to automatically create yaml files (given a base config) for e.g. hyper parameter tuning. Is there anything else you would like to see? Very happy for feedback!
- I agree on the tests. We have a couple of tests and they also run on every PR/push automatically via Github actions but we definitely should extend our test coverage (it is a TODO on our list...)
- Good suggestions, thanks. We did some related studies in the past (on explanability) but never integrated the specific functions into the library (but instead published them in a separate repository with the paper code). I agree that it makes totally sense to include such utilities also into the library. When we started, our vision for NeuralHydrology was more about doing quick experiments with different model architectures/datasets/settings and the entire part of the post-training analysis was always separated. Regarding uncertainty: This is actually something that is well integrated into the library already and we also published a paper about it. We have a couple of different options that one can use: Different mixture density heads ([GMM](https://neuralhydrology.readthedocs.io/en/latest/usage/models.html#gmm), [CMAL](https://neuralhydrology.readthedocs.io/en/latest/usage/models.html#cmal), [UMAL](https://neuralhydrology.readthedocs.io/en/latest/usage/models.html#umal)) that all return probabilistic outputs (instead of a point estimate in the standard regression setting) and we also support MC-Dropout sampling during evaluation (see `mc_dropout` under the [in the list of config arguments](https://neuralhydrology.readthedocs.io/en/latest/usage/config.html)). We also plan to release a tutorial on all of these options, as soon as the [corresponding paper](https://hess.copernicus.org/preprints/hess-2021-154/) is accepted for publication.

Again, thank you so much for your time reviewing our library/paper and for your feedback so far!

**Comment by kratzert** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.04050/joss.04050/10.21105.joss.04050.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.04050/joss.04050/10.21105.joss.04050.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
OK, the reviewer has been re-invited.

@chuckaustin please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by elbeejay** [comment]
Hello all :wave:

I wanted to thank our reviewers for testing and evaluating this software. To hopefully help @kratzert and co-authors make changes, I wanted to ask @jhamman to provide some clarification on the unchecked boxes in the documentation and software paper sections. Please either open issues in the repositories or make a small comment here if these are concerns about missing documentation, tests, or community guidelines. I believe the paper has been revised since your review to try and address some of your concerns.

Similarly @chuckaustin, if you could check off the boxes corresponding to things you believe the authors have accomplished, this would help narrow the scope of the revisions you'd like to see made, and complement the open issues.

Again thanks all, @kratzert please let me or any of the reviewers know if you need further clarification about open issues or unchecked items on reviewer checklists.

Thanks,
Jay

**Comment by kratzert** [comment]
Hi @elbeejay 

sorry for my delayed reply. I was on a business trip and hadn't had much time to look into it in the last days. I just checked all remaining review comments in our repository and I think they are pretty clear. 

- I think I could solve the installation issues of @chuckaustin but I am waiting for his confirmation (see https://github.com/neuralhydrology/neuralhydrology/issues/76#issuecomment-1054656670). We will remove the CUDA9.2 environment file and we will add one for cuda11.3 (which, together with 10.2, are the two supported CUDA versions for the stable PyTorch installation).
- I will also update the `GenericDataset` description in the tutorial, but I am waiting for feedback of @chuckaustin before committing, see https://github.com/neuralhydrology/neuralhydrology/issues/77#issuecomment-1054636471

Other than that, I think there is nothing left open for me/us to do, right? I am not sure about the unchecked checkboxes above though.

**Comment by elbeejay** [comment]
Sounds good @kratzert, it looks like the installation and example checkboxes from @chuckaustin's review are all that remain. It seems like the path to resolving the two outstanding issues is relatively clear which is good to hear. You are correct that there is nothing else for you and your co-authors to do at this time.

Once the outstanding issues are resolved, I'll do a final check of the submitted materials, ask you to create an archive and tagged release for the reviewed code, and that's about it.

Thanks,
Jay

**Comment by elbeejay** [comment]
Hi @kratzert, it looks like those last few issues have been resolved. I wanted to follow up on some of the discussion that happened in [issue 77](https://github.com/neuralhydrology/neuralhydrology/issues/77) - it sounded like you and @chuckaustin agreed that it would be useful to expand the documentation in response to some of the questions raised in that issue. Do you plan on making those changes to the documentation prior to publication? I see that the issue was closed, so this is by no means a requirement for publication, but I figured I would ask first.

Thanks

**Comment by elbeejay** [comment]
@editorialbot check references

**Comment by elbeejay** [comment]
@editorialbot set paper as branch

**Comment by elbeejay** [comment]
@editorialbot check references

**Comment by elbeejay** [comment]
@editorialbot generate pdf

**Comment by elbeejay** [comment]
@kratzert one minor suggestion for the paper text:
- Line 19: Switch the colon to a period, or remove the capital "P" in "Pre-built" on the following line

Besides this, I think things look good. I am optimistic that this package will help accelerate deep learning acceptance/adoption in the hydrology community. After making the minor revision noted above, I need you to do a few things at this time:

- [ ] Make a tagged release of your software, and list the version tag of the archived version here.
- [ ] Archive the reviewed software in Zenodo or a similar service (e.g., figshare, an institutional repository)
- [ ] Check the archival deposit (e.g., in Zenodo) has the correct metadata. This includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it). You may also add the authors' ORCID.
- [ ] Please list the DOI of the archived version here.

Please let me know when these steps are complete, they allow us to finalize the metadata that will be associated with the JOSS publication. Once that is done I can recommend that we accept and publish this paper.

**Comment by kratzert** [comment]
> Hi @kratzert, it looks like those last few issues have been resolved. I wanted to follow up on some of the discussion that happened in [issue 77](https://github.com/neuralhydrology/neuralhydrology/issues/77) - it sounded like you and @chuckaustin agreed that it would be useful to expand the documentation in response to some of the questions raised in that issue. Do you plan on making those changes to the documentation prior to publication? I see that the issue was closed, so this is by no means a requirement for publication, but I figured I would ask first.
> 
> Thanks

Hi @elbeejay,

I will try to make the changes today yes, definitely plan to add them prior to the publication (tagged release), since it is not a lot of work. Also thanks for the typo, I will make sure that is fixed. I will come back to you as soon as I have everything done.

**Comment by kratzert** [comment]
> @kratzert one minor suggestion for the paper text:
> 
>     * Line 19: Switch the colon to a period, or remove the capital "P" in "Pre-built" on the following line
> 
> 
> Besides this, I think things look good. I am optimistic that this package will help accelerate deep learning acceptance/adoption in the hydrology community. After making the minor revision noted above, I need you to do a few things at this time:
> 
>     * [ ]  Make a tagged release of your software, and list the version tag of the archived version here.
> 
>     * [ ]  Archive the reviewed software in Zenodo or a similar service (e.g., figshare, an institutional repository)
> 
>     * [ ]  Check the archival deposit (e.g., in Zenodo) has the correct metadata. This includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it). You may also add the authors' ORCID.
> 
>     * [ ]  Please list the DOI of the archived version here.
> 
> 
> Please let me know when these steps are complete, they allow us to finalize the metadata that will be associated with the JOSS publication. Once that is done I can recommend that we accept and publish this paper.

- Link to tagged version: https://github.com/neuralhydrology/neuralhydrology/releases/tag/v.1.2.3
- Link to archived software: https://zenodo.org/record/6326394
- Checked all authors names, added ORCID ids
- DOI: https://doi.org/10.5281/zenodo.6326394

The paper is also updated with the typo fix.

Anything else you need?

**Comment by elbeejay** [comment]
@editorialbot set v.1.2.3 as version

**Comment by elbeejay** [comment]
@editorialbot set 10.5281/zenodo.6326394 as archive

**Comment by elbeejay** [comment]
> Anything else you need?

@kratzert that is it, thank you for taking care of those last items. I wanted to thank the folks that peer-reviewed this submission once again as well - so thank you to @jhamman, @ammilten, and @chuckaustin. 

At this time I will be recommending that we accept and publish this submission in JOSS. The paper and metadata will get one final review by an Editor-in-Chief prior to publication.

**Comment by elbeejay** [comment]
@editorialbot recommend-accept

**Comment by kratzert** [comment]
> > Anything else you need?
> 
> @kratzert that is it, thank you for taking care of those last items. I wanted to thank the folks that peer-reviewed this submission once again as well - so thank you to @jhamman, @ammilten, and @chuckaustin.

Also thank you from my side for taking care of the entire review process @elbeejay and  @jhamman, @ammilten and @chuckaustin  for reviewing our package and providing useful feedback!

> At this time I will be recommending that we accept and publish this submission in JOSS. The paper and metadata will get one final review by an Editor-in-Chief prior to publication.

Juhu :tada:

**Comment by kyleniemeyer** [comment]
Hi @kratzert, I made a small PR with some formatting fixes to the paper: https://github.com/neuralhydrology/neuralhydrology/pull/79

Could you merge this? Then I can publish the paper.

**Comment by kratzert** [comment]
> Hi @kratzert, I made a small PR with some formatting fixes to the paper: [neuralhydrology/neuralhydrology#79](https://github.com/neuralhydrology/neuralhydrology/pull/79)
> 
> Could you merge this? Then I can publish the paper.

Done. Thanks for the fixes.

**Comment by kyleniemeyer** [comment]
@editorialbot accept

**Comment by kyleniemeyer** [comment]
Congratulations @kratzert on your article's publication in JOSS!
Many thanks to @ammilten, @chuckaustin, and @jhamman for reviewing this, and @elbeejay for editing.
